Zed cannot be opened even if `ZED_DEVICE_ID` is set in a shell.

### Summary

When the `v0.181` is released, I did export `ZED_DEVICE_ID` to run Zed with the NVIDIA GPU. I followed procedures to do this (https://github.com/zed-industries/zed/releases/tag/v0.181.5 -> Linux-video section). 

For the moment, my solution is to edit the desktop entry app like this:

```
[Desktop Entry]
Version=1.0
Type=Application
Name=Zed
GenericName=Text Editor
Comment=A high-performance, multiplayer code editor.
TryExec=/home/tamdaz/.local/zed.app/bin/zed
StartupNotify=true
Exec=env __NV_PRIME_RENDER_OFFLOAD=1 __GLX_VENDOR_LIBRARY_NAME=nvidia /home/tamdaz/.local/zed.app/bin/zed %U
Icon=/home/tamdaz/.local/zed.app/share/icons/hicolor/512x512/apps/zed.png
Categories=Utility;TextEditor;Development;IDE;
Keywords=zed;
MimeType=text/plain;application/x-zerosize;x-scheme-handler/zed;
Actions=NewWorkspace;

[Desktop Action NewWorkspace]
Exec=env __NV_PRIME_RENDER_OFFLOAD=1 __GLX_VENDOR_LIBRARY_NAME=nvidia /home/tamdaz/.local/zed.app/bin/zed --new %U
Name=Open a new workspace
```

where I added environment variables to start Zed with Nvidia.

```
env __NV_PRIME_RENDER_OFFLOAD=1 __GLX_VENDOR_LIBRARY_NAME=nvidia
```

### Description
Steps to trigger the problem:
1. Export the `ZED_DEVICE_ID` with the GPU ID (thanks to lspci).
2. Running `zed` command from the terminal (I currently use `zsh`).

Actual Behavior:

Zed is immediately closed after running the `zed` command.

Output shell:
```
[2025-04-10T21:55:26+02:00 ERROR gpui] Failed to parse device ID from `ZED_DEVICE_ID` environment variable

Caused by:
    invalid digit found in string
```

Expected Behavior:

It should be opened even if `ZED_DEVICE_ID` variable is exported.

### Zed Version and System Specs

App version: 0.181.5

```
ZED_DEVICE_ID=0x1f91 zed --foreground
xkbcommon: ERROR: couldn't find a Compose file for locale "C"
[2025-04-10T21:55:26+02:00 ERROR gpui] Failed to parse device ID from `ZED_DEVICE_ID` environment variable

Caused by:
    invalid digit found in string
[2025-04-10T21:55:27+02:00 ERROR zed::reliability] {
  "thread": "main",
  "payload": "GPU error ERROR_SURFACE_LOST_KHR",
  "location_data": {
    "file": "/home/ubuntu/.cargo/git/checkouts/blade-26bd638d88226de3/b16f5c7/blade-graphics/src/vulkan/command.rs",
    "line": 455
  },
  "backtrace": [
    "zed::reliability::init_panic_hook::{{closure}}::hd6bae70565b70b32+140697704",
    "std::panicking::rust_panic_with_hook::h66e909d048c263a9+114671651",
    "std::panicking::begin_panic_handler::{{closure}}::h8d9aa8be7e8634cf+114670874",
    "std::sys::backtrace::__rust_end_short_backtrace::h7d7e47ef99abf6aa+114664841",
    "rust_begin_unwind+114670013",
    "core::panicking::panic_fmt::hf8ffc7c15bfb58a0+8644928",
    "<blade_graphics::hal::Context as blade_graphics::traits::CommandDevice>::submit::h2de7ff55004fe10b+30923424",
    "gpui::platform::blade::blade_renderer::BladeRenderer::draw::h5015346e76bb5699+67746261",
    "<gpui::platform::linux::x11::window::X11Window as gpui::platform::PlatformWindow>::draw::h9842f10bc79f5e25+67682444",
    "gpui::window::Window::new::{{closure}}::h41c637426afa9b5f+67894188",
    "gpui::platform::linux::x11::client::X11Client::process_x11_events::h8d51a99b049c6071+67579748",
    "<core::cell::RefCell<calloop::sources::DispatcherInner<S,F>> as calloop::sources::EventDispatcher<Data>>::process_events::h5a3c6493e13f3099+64247471",
    "<gpui::platform::linux::x11::client::X11Client as gpui::platform::linux::platform::LinuxClient>::run::hfd8b88654886c35e+67644432",
    "gpui::platform::linux::platform::<impl gpui::platform::Platform for P>::run::h092aa00069327c70+67246605",
    "gpui::app::Application::run::h144b61bb13898649+138173929",
    "zed::main::h50142b7475dc2f58+140726055",
    "std::sys::backtrace::__rust_begin_short_backtrace::h629ba96cd76cc8ec+141241411",
    "std::rt::lang_start::{{closure}}::h4aaf4e80648e2818+141473401",
    "std::rt::lang_start_internal::heee0af441e41a6d2+114595453",
    "main+140742524",
    "__libc_start_call_main+45853649707594",
    "__libc_start_main_impl+45853649707781",
    "_start+10906382"
  ],
  "app_version": "0.181.5",
  "app_commit_sha": "4f04f932c1e6a296acde3f93e078d23f1e6e44ca",
  "release_channel": "stable",
  "target": "x86_64-unknown-linux-gnu",
  "os_name": "Linux X11",
  "os_version": "debian 12",
  "architecture": "x86_64",
  "panicked_on": 1744314927291,
  "system_id": "eec50b3c-6b01-4313-ab3f-b8fa62b77bb0",
  "installation_id": "c45f5afb-5eba-4d20-a7e7-550401d25989",
  "session_id": "43d111df-7491-4f4b-b700-e146f1f1b9f3"
}
```

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary>

```log
2025-04-10T21:52:19.862533935+02:00 [INFO] ========== starting zed ==========
2025-04-10T21:52:19.885452146+02:00 [INFO] Compositor detection: _NET_WM_CM_S?=false, _NET_WM_CM_OWNER=false, _NET_SUPPORTING_WM_CHECK=true
2025-04-10T21:52:19.885509638+02:00 [INFO] x11: compositor present: true, gtk_frame_extents_supported: true
2025-04-10T21:52:19.886977313+02:00 [ERROR] Failed to parse device ID from `ZED_DEVICE_ID` environment variable

Caused by:
    invalid digit found in string
2025-04-10T21:52:19.914087371+02:00 [INFO] Presentation support: "VK_KHR_wayland_surface"
2025-04-10T21:52:19.91412395+02:00 [INFO] Presentation support: "VK_KHR_xcb_surface"
2025-04-10T21:52:19.914146706+02:00 [INFO] Presentation support: "VK_KHR_xlib_surface"
2025-04-10T21:52:19.914172129+02:00 [INFO] Enabling Vulkan Portability
2025-04-10T21:52:19.926522608+02:00 [INFO] Adapter: "Intel(R) UHD Graphics 630 (CFL GT2)"
2025-04-10T21:52:19.926685245+02:00 [INFO] No ray tracing extensions are supported
2025-04-10T21:52:19.928379077+02:00 [INFO] Try connect ibus
2025-04-10T21:52:19.928881705+02:00 [INFO] perform;
2025-04-10T21:52:19.928911948+02:00 [INFO] authenticate;
2025-04-10T21:52:19.928948469+02:00 [INFO] read_command;
2025-04-10T21:52:19.929026794+02:00 [INFO] Opening main db
2025-04-10T21:52:19.929096599+02:00 [INFO] send_secondary_commands;
2025-04-10T21:52:19.929619451+02:00 [INFO] new;
2025-04-10T21:52:19.929649592+02:00 [INFO] socket reader;
2025-04-10T21:52:19.929674866+02:00 [INFO] read_socket;
2025-04-10T21:52:19.929711771+02:00 [INFO] read_socket;
2025-04-10T21:52:19.929825905+02:00 [INFO] read_socket;
2025-04-10T21:52:19.929852159+02:00 [INFO] Opening main db
2025-04-10T21:52:19.929946025+02:00 [INFO] read_socket;
2025-04-10T21:52:19.930044784+02:00 [INFO] read_socket;
2025-04-10T21:52:19.930255383+02:00 [INFO] read_socket;
2025-04-10T21:52:19.930294798+02:00 [INFO] keep_updated;
2025-04-10T21:52:19.931310222+02:00 [INFO] read_socket;
2025-04-10T21:52:19.931384089+02:00 [INFO] Using git binary path: None
2025-04-10T21:52:19.932019596+02:00 [INFO] read_socket;
2025-04-10T21:52:19.932548973+02:00 [INFO] read_socket;
2025-04-10T21:52:19.932680198+02:00 [INFO] read_socket;
2025-04-10T21:52:19.932800793+02:00 [INFO] read_socket;
2025-04-10T21:52:19.932913695+02:00 [INFO] read_socket;
2025-04-10T21:52:19.932992639+02:00 [INFO] read_socket;
2025-04-10T21:52:19.933131111+02:00 [INFO] read_socket;
2025-04-10T21:52:19.933231594+02:00 [INFO] read_socket;
2025-04-10T21:52:19.952545126+02:00 [INFO] No prompt template overrides directory found at /home/tamdaz/.local/share/zed/prompt_overrides. Using built-in prompts.
2025-04-10T21:52:19.970608019+02:00 [INFO] extensions updated. loading 15, reloading 0, unloading 0
2025-04-10T21:52:19.971899409+02:00 [INFO] activate is not implemented on Linux, ignoring the call
2025-04-10T21:52:19.973386401+02:00 [INFO] Opening main db
2025-04-10T21:52:19.973721019+02:00 [INFO] perform;
2025-04-10T21:52:19.973758594+02:00 [INFO] authenticate;
2025-04-10T21:52:19.97381214+02:00 [INFO] read_command;
2025-04-10T21:52:19.973857368+02:00 [INFO] send_secondary_commands;
2025-04-10T21:52:19.974497914+02:00 [INFO] socket reader;
2025-04-10T21:52:19.974539145+02:00 [INFO] read_socket;
2025-04-10T21:52:19.974595255+02:00 [INFO] read_socket;
2025-04-10T21:52:19.980789542+02:00 [INFO] read_socket;
2025-04-10T21:52:19.982667935+02:00 [INFO] read_socket;
2025-04-10T21:52:19.982975239+02:00 [INFO] read_socket;
2025-04-10T21:52:19.983321164+02:00 [INFO] read_socket;
2025-04-10T21:52:19.983361827+02:00 [INFO] new;
2025-04-10T21:52:19.983545574+02:00 [INFO] read_socket;
2025-04-10T21:52:19.983686846+02:00 [INFO] read_socket;
2025-04-10T21:52:19.98381728+02:00 [INFO] read_socket;
2025-04-10T21:52:19.98432269+02:00 [INFO] read_socket;
2025-04-10T21:52:19.984396271+02:00 [INFO] keep_updated;
2025-04-10T21:52:19.984704779+02:00 [INFO] read_socket;
2025-04-10T21:52:19.984853508+02:00 [INFO] read_socket;
2025-04-10T21:52:19.984991079+02:00 [INFO] read_socket;
2025-04-10T21:52:20.006405188+02:00 [INFO] Using Visual { id: 35, colormap: 0, depth: 32 }
2025-04-10T21:52:20.006483112+02:00 [INFO] Creating colormap 79691779
2025-04-10T21:52:20.007226063+02:00 [INFO] Using surface present mode MAILBOX
2025-04-10T21:52:20.009796223+02:00 [INFO] set environment variables from shell:/usr/bin/zsh, path:/home/tamdaz/.config/composer/vendor/bin/:/home/tamdaz/bin:/usr/local/bin:/home/tamdaz/.config/composer/vendor/bin/:/home/tamdaz/bin:/usr/local/bin:/home/tamdaz/.cargo/bin:/home/tamdaz/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/home/tamdaz/.dotnet/tools:/home/tamdaz/.config/composer/vendor/bin:/home/tamdaz/.cargo/bin:/home/tamdaz/.config/composer/vendor/bin:/home/tamdaz/.cargo/bin
2025-04-10T21:52:20.014872748+02:00 [INFO] Creating a descriptor pool for at most 16 sets
2025-04-10T21:52:20.01494737+02:00 [INFO] Creating a descriptor pool for at most 16 sets
2025-04-10T21:52:20.014985154+02:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: Ignored }
2025-04-10T21:52:20.120726546+02:00 [INFO] Refreshing at 5556 micros
2025-04-10T21:52:20.122027608+02:00 [INFO] Using GPU: GpuSpecs { is_software_emulated: false, device_name: "Intel(R) UHD Graphics 630 (CFL GT2)", driver_name: "Intel open-source Mesa driver", driver_info: "Mesa 22.3.6" }
2025-04-10T21:52:20.14713587+02:00 [INFO] set status on client 0: Authenticating
2025-04-10T21:52:20.147381139+02:00 [INFO] perform;
2025-04-10T21:52:20.147417574+02:00 [INFO] authenticate;
2025-04-10T21:52:20.147457701+02:00 [INFO] read_command;
2025-04-10T21:52:20.147524586+02:00 [INFO] send_secondary_commands;
2025-04-10T21:52:20.148043996+02:00 [INFO] socket reader;
2025-04-10T21:52:20.148073455+02:00 [INFO] read_socket;
2025-04-10T21:52:20.148111401+02:00 [INFO] read_socket;
2025-04-10T21:52:20.153480662+02:00 [INFO] read_socket;
2025-04-10T21:52:20.154541305+02:00 [INFO] read_socket;
2025-04-10T21:52:20.154836741+02:00 [INFO] read_socket;
2025-04-10T21:52:20.155272827+02:00 [INFO] read_socket;
2025-04-10T21:52:20.155312377+02:00 [INFO] new;
2025-04-10T21:52:20.155512702+02:00 [INFO] read_socket;
2025-04-10T21:52:20.155703439+02:00 [INFO] read_socket;
2025-04-10T21:52:20.156199497+02:00 [INFO] read_socket;
2025-04-10T21:52:20.156794514+02:00 [INFO] read_socket;
2025-04-10T21:52:20.156879694+02:00 [INFO] keep_updated;
2025-04-10T21:52:20.157341068+02:00 [INFO] read_socket;
2025-04-10T21:52:20.15756926+02:00 [INFO] read_socket;
2025-04-10T21:52:20.157687178+02:00 [INFO] read_socket;
2025-04-10T21:52:20.165475281+02:00 [INFO] Using surface present mode MAILBOX
2025-04-10T21:52:20.165513419+02:00 [INFO] Retaining current format: Bgra8UnormSrgb
2025-04-10T21:52:20.185790946+02:00 [INFO] Using surface present mode MAILBOX
2025-04-10T21:52:20.185826503+02:00 [INFO] Retaining current format: Bgra8UnormSrgb
2025-04-10T21:52:20.194598857+02:00 [INFO] XConnected server on 33554463, transport version: 0.0, TRANSPORT_MAX: 20
2025-04-10T21:52:20.207923794+02:00 [INFO] Opening main db
2025-04-10T21:52:20.208239283+02:00 [INFO] set status on client 298729: Connecting
2025-04-10T21:52:20.228162248+02:00 [INFO] connected to rpc endpoint https://collab.zed.dev/rpc
2025-04-10T21:52:20.312548593+02:00 [ERROR] {
  "thread": "main",
  "payload": "Aquire image error ERROR_SURFACE_LOST_KHR",
  "location_data": {
    "file": "/home/ubuntu/.cargo/git/checkouts/blade-26bd638d88226de3/b16f5c7/blade-graphics/src/vulkan/surface.rs",
    "line": 50
  },
  "backtrace": [
    "zed::reliability::init_panic_hook::{{closure}}::hd6bae70565b70b32+140697704",
    "std::panicking::rust_panic_with_hook::h66e909d048c263a9+114671651",
    "std::panicking::begin_panic_handler::{{closure}}::h8d9aa8be7e8634cf+114670874",
    "std::sys::backtrace::__rust_end_short_backtrace::h7d7e47ef99abf6aa+114664841",
    "rust_begin_unwind+114670013",
    "core::panicking::panic_fmt::hf8ffc7c15bfb58a0+8644928",
    "blade_graphics::hal::surface::<impl blade_graphics::hal::Surface>::acquire_frame::h777a67a00fcef116+30865502",
    "gpui::platform::blade::blade_renderer::BladeRenderer::draw::h5015346e76bb5699+67740299",
    "<gpui::platform::linux::x11::window::X11Window as gpui::platform::PlatformWindow>::draw::h9842f10bc79f5e25+67682444",
    "gpui::window::Window::new::{{closure}}::{{closure}}::h0014f7459f448c6f+67902919",
    "gpui::window::Window::new::{{closure}}::h41c637426afa9b5f+67890747",
    "<core::cell::RefCell<calloop::sources::DispatcherInner<S,F>> as calloop::sources::EventDispatcher<Data>>::process_events::h5a3c6493e13f3099+64247418",
    "<gpui::platform::linux::x11::client::X11Client as gpui::platform::linux::platform::LinuxClient>::run::hfd8b88654886c35e+67644432",
    "gpui::platform::linux::platform::<impl gpui::platform::Platform for P>::run::h092aa00069327c70+67246605",
    "gpui::app::Application::run::h144b61bb13898649+138173929",
    "zed::main::h50142b7475dc2f58+140726055",
    "std::sys::backtrace::__rust_begin_short_backtrace::h629ba96cd76cc8ec+141241411",
    "std::rt::lang_start::{{closure}}::h4aaf4e80648e2818+141473401",
    "std::rt::lang_start_internal::heee0af441e41a6d2+114595453",
    "main+140742524",
    "__libc_start_call_main+46200903516746",
    "__libc_start_main_impl+46200903516933",
    "_start+10906382"
  ],
  "app_version": "0.181.5",
  "app_commit_sha": "4f04f932c1e6a296acde3f93e078d23f1e6e44ca",
  "release_channel": "stable",
  "target": "x86_64-unknown-linux-gnu",
  "os_name": "Linux X11",
  "os_version": "debian 12",
  "architecture": "x86_64",
  "panicked_on": 1744314740312,
  "system_id": "eec50b3c-6b01-4313-ab3f-b8fa62b77bb0",
  "installation_id": "c45f5afb-5eba-4d20-a7e7-550401d25989",
  "session_id": "ebb1a9ae-de8d-492d-a907-65262b7ac741"
}

```

</details>

--- a/crates/gpui/src/platform/blade/blade_context.rs
+++ b/crates/gpui/src/platform/blade/blade_context.rs
@@ -1,5 +1,6 @@
 use anyhow::Context as _;
 use blade_graphics as gpu;
+use itertools::Itertools;
 use std::sync::Arc;
 use util::ResultExt;
 
@@ -11,8 +12,7 @@ pub struct BladeContext {
 impl BladeContext {
     pub fn new() -> anyhow::Result<Self> {
         let device_id_forced = match std::env::var("ZED_DEVICE_ID") {
-            Ok(val) => val
-                .parse()
+            Ok(val) => parse_pci_id(&val)
                 .context("Failed to parse device ID from `ZED_DEVICE_ID` environment variable")
                 .log_err(),
             Err(std::env::VarError::NotPresent) => None,
@@ -34,5 +34,49 @@ impl BladeContext {
             .map_err(|e| anyhow::anyhow!("{:?}", e))?,
         );
         Ok(Self { gpu })
+    }
+}
+
+fn parse_pci_id(id: &str) -> anyhow::Result<u32> {
+    let mut id = id.trim();
+
+    if id.starts_with("0x") || id.starts_with("0X") {
+        id = &id[2..];
+    }
+    let is_hex_string = id.chars().all(|c| c.is_ascii_hexdigit());
+    let is_4_chars = id.len() == 4;
+    anyhow::ensure!(
+        is_4_chars && is_hex_string,
+        "Expected a 4 digit PCI ID in hexadecimal format"
+    );
+
+    return u32::from_str_radix(id, 16)
+        .map_err(|_| anyhow::anyhow!("Failed to parse PCI ID as hex"));
+}
+
+#[cfg(test)]
+mod tests {
+    use super::parse_pci_id;
+
+    #[test]
+    fn test_parse_device_id() {
+        assert!(parse_pci_id("0xABCD").is_ok());
+        assert!(parse_pci_id("ABCD").is_ok());
+        assert!(parse_pci_id("abcd").is_ok());
+        assert!(parse_pci_id("1234").is_ok());
+        assert!(parse_pci_id("123").is_err());
+        assert_eq!(
+            parse_pci_id(&format!("{:x}", 0x1234)).unwrap(),
+            parse_pci_id(&format!("{:X}", 0x1234)).unwrap(),
+        );
+
+        assert_eq!(
+            parse_pci_id(&format!("{:#x}", 0x1234)).unwrap(),
+            parse_pci_id(&format!("{:#X}", 0x1234)).unwrap(),
+        );
+        assert_eq!(
+            parse_pci_id(&format!("{:#x}", 0x1234)).unwrap(),
+            parse_pci_id(&format!("{:#X}", 0x1234)).unwrap(),
+        );
     }
 }

--- a/crates/gpui/src/platform/blade/blade_context.rs
+++ b/crates/gpui/src/platform/blade/blade_context.rs
@@ -1,6 +1,5 @@
 use anyhow::Context as _;
 use blade_graphics as gpu;
-use itertools::Itertools;
 use std::sync::Arc;
 use util::ResultExt;
 

